---
title: '#4 Calculation of microbiome variance explained by phenotypes including main
  Figure 4'
output: html_document
---

# ======================================================================
# By: Weersma Group, TransplantLines (UMCG)
#
# UMCG Transplantlines Project data analysis
# ======================================================================
#   Calculation of microbiome variance explained by phenotypes -
#
# These codes perform microbiome-variance-explained (ADONIS) analysis
# for Liver transplant recipients and Renal transplant recipient
#  (results included in the Fig 4)
#
# ======================================================================
#
# NOTE: This is implementation on Mock Data published in this github repo,
#       and is intended for demonstration purposes only, the
#       results are not identical to Figures in the manuscript.
#       Analysis is implemented on lower number of permutations 
#       for quicker execution (999 instead of 9999 as used in the manuscript)
#
# NOTE2: Codes are intended to be run from root of this github repo,
#       if running them from different location, make sure to adjust setwd
#       to appropriate location (root of this github repo)
#
# ======================================================================
#
# make sure following packages are installed:
# install.packages(c('dplyr','foreign','vegan','ggplot2'))
#
# ======================================================================

```{r}
#Load extra functions for microbiome processing 
source('r_scripts_library/R_Microbiome_scripts.R')
library(dplyr)
library(foreign)
library(vegan)
library(ggplot2)
```
```{r}
#ADONIS - Within Liver transplant recipients ----
#Load Data
CDtaxa <- read.table("mock_data/Mocks_Taxa_Filtered.txt",sep='\t',header=T, fill=T)
CDtaxa$ID <- gsub('MOCK_','',trimws(as.character(CDtaxa$ID)))
rownames(CDtaxa) <- CDtaxa$ID

CDmeta <- read.csv("mock_data/MOCK_Meta_cleaned_v2.csv", row.names=1)
CDmeta$ID <- row.names(CDmeta)
CDmeta$ID <- trimws(CDmeta$ID)
row.names(CDmeta) <- CDmeta$ID
```
```{r}
# Liver Transplant Recipients Data
# Get liver transplants only
ADONIS_Within_LTxOld <- subset(CDmeta, Cross_sectional_samples =="LTR")
ADONIS_Within_LTxOld_Taxa_t <- ADONIS_Within_LTxOld[c("ID")]

inDF <- CDtaxa
# Filter microbiome, keep only species
spDF <- filterMetaGenomeDF(inDF=subsetMicrobiomeDF(inDF,getTaxa = T,getPWYs = F,getCARDs = F,getVFs = F,getDivs = F,getPhenos = F,verbose = T),
                           minMRelAb = 0,presPerc = 0,minMedRelAb = 0,keepLevels = "S")
#Prep datasets: Taxa
ADONIS_Within_LTxOld_Taxa <- merge(ADONIS_Within_LTxOld_Taxa_t, spDF, by="row.names")
rownames(ADONIS_Within_LTxOld_Taxa) <- ADONIS_Within_LTxOld_Taxa$ID
#Prep datasets: Metadata
ADONIS_Within_LTxOld_meta <- ADONIS_Within_LTxOld_Taxa[c("ID")]
rownames(ADONIS_Within_LTxOld_meta ) <- ADONIS_Within_LTxOld_meta$ID
ADONIS_Within_LTxOld_meta <- merge(ADONIS_Within_LTxOld_meta, CDmeta, by="row.names")
rownames(ADONIS_Within_LTxOld_meta) <- ADONIS_Within_LTxOld_meta$ID.x
#Clean datasets
ADONIS_Within_LTxOld_Taxa$ID <- NULL
ADONIS_Within_LTxOld_Taxa$Row.names <- NULL
ADONIS_Within_LTxOld_meta$ID <- ADONIS_Within_LTxOld_meta$ID.x

#Select vars to use
adonisVarsTouse <- select(ADONIS_Within_LTxOld_meta,
                          "Age",
                          "Gender",
                          "Years_since_transplantation",
                          "BMI",
                          "BSA",
                          "BIAWAISTHIP",
                          "BIAFATPC",
                          "SBP",
                          "Proteinuria",
                          "Retransplantation",
                          "func_constipation",
                          "IBS",
                          "func_diarrhea",
                          "Hypertension",
                          "Alcohol",
                          "Smoking",
                          "Diabetes",
                          "GGT",
                          "ALT",
                          "AST",
                          "ALP",
                          "Tbil",
                          "CRP",
                          "Leukocytes",
                          "eGFR",
                          "HDL",
                          "LDL",
                          "Total_cholesterol",
                          "NT_proBNP",
                          "Troponin_T",
                          "PPI",
                          "Statins",
                          "Laxative",
                          "Antibiotics",
                          "Mycophenolic_acid",
                          "Azathioprine",
                          "Ciclosporine",
                          "Tacrolimus",
                          "Prednisolone",
                          "Sirolimus",
                          "Tac_Pred",
                          "Cyclo_Pred",
                          "MMF_Cyclo",
                          "MMF_Pred",
                          "MMF_Tac",
                          "MMF_Cyclo_Pred",
                          "MMF_Tac_Pred",
                          "Aza_Tac_Pred")

# UNIVARIATE ADONIS
rownames(adonisVarsTouse) <- rownames(ADONIS_Within_LTxOld_meta)
# do univariate adonis
adonis_meta <- matrix(ncol = 7, nrow=ncol(adonisVarsTouse))
for (i in 1:ncol(adonisVarsTouse)){
  varToUse <- colnames(adonisVarsTouse)[[i]]
  print (paste(' >> grinding ',colnames(adonisVarsTouse)[[i]]))
  # make new DF with selected phenotype and taxa 
  tmpDF <- merge.data.frame(ADONIS_Within_LTxOld_meta[,colnames(ADONIS_Within_LTxOld_meta) %in% c("ID",varToUse)] ,
                            spDF,by.x="ID",by.y = "row.names")
  # get rid of ID (used only for merging)
  tmpDF$ID <- NULL
  # get rid of rows with NAs
  #print('debug, DF size before removing NAs:')
  #print(dim(tmpDF))
  tmpDF <- tmpDF[complete.cases(tmpDF),]
  #print('debug, DF size after removing NAs:')
  #print(dim(tmpDF))
  # calculate BC matrix
  tmpDFforBC <- tmpDF
  tmpDFforBC[[varToUse]] <- NULL
  print('calculating BC matrix')
  bcMat <- vegdist(tmpDFforBC,method = "bray")
  print(paste0('doing adonis for ',varToUse))
  ad<-adonis(bcMat ~ adonisVarsTouse[,i],permutations=999)
  aov_table <- ad$aov.tab
  #Df
  adonis_meta[i,1]=aov_table[1,1]
  #SumsOfSqs
  adonis_meta[i,2]=aov_table[1,2]
  #MeanSqs
  adonis_meta[i,3]=aov_table[1,3]
  #FModel
  adonis_meta[i,4]=aov_table[1,4]
  #R2
  adonis_meta[i,5]=aov_table[1,5]
  #Pval
  adonis_meta[i,6]=aov_table[1,6]
}

adonis_meta= as.data.frame(adonis_meta)
adonis_meta$V7=p.adjust(adonis_meta$V6, method = "BH")
adonis_meta$V8="No"
adonis_meta$V8[adonis_meta$V7<0.05]="Yes"
colnames(adonis_meta)=c("DF", "SumsOfSqs", "MeanSqs", "FModel","R2","pval","FDR(BH)","Significant")
rownames(adonis_meta) = colnames(adonisVarsTouse)
```
```{r}
# View(adonis_meta)
g <- ggplot(adonis_meta, aes(reorder(row.names(adonis_meta), R2), R2, fill=Significant)) + 
  geom_bar(stat = "identity") + coord_flip() + theme_bw() + 
  ylab ("Explained variance (R^2) ") + xlab ("Factor")  + theme(text = element_text(size=11))
plot(g)
```
```{r}
#ADONIS - Within Renal transplant recipients---- 
ADONIS_Within_NTxOld <- subset(CDmeta, Cross_sectional_samples =="RTR")
ADONIS_Within_NTxOld_Taxa <- ADONIS_Within_NTxOld[c("ID")]
inDF <- CDtaxa
spDF <- filterMetaGenomeDF(inDF=subsetMicrobiomeDF(inDF,getTaxa = T,getPWYs = F,getCARDs = F,getVFs = F,getDivs = F,getPhenos = F,verbose = T),
                           minMRelAb = 0,presPerc = 0,minMedRelAb = 0,keepLevels = "S")
ADONIS_Within_NTxOld_Taxa <- merge(ADONIS_Within_NTxOld_Taxa, spDF, by="row.names")
rownames(ADONIS_Within_NTxOld_Taxa) <- ADONIS_Within_NTxOld_Taxa$ID
ADONIS_Within_NTxOld_meta <- ADONIS_Within_NTxOld_Taxa[c("ID")]
rownames(ADONIS_Within_NTxOld_meta ) <- ADONIS_Within_NTxOld_meta$ID
ADONIS_Within_NTxOld_meta <- merge(ADONIS_Within_NTxOld_meta, CDmeta, by="row.names")
rownames(ADONIS_Within_NTxOld_meta) <- ADONIS_Within_NTxOld_meta$ID.x
ADONIS_Within_NTxOld_Taxa$ID <- NULL
ADONIS_Within_NTxOld_Taxa$Row.names <- NULL

adonisVarsTouse <- select(ADONIS_Within_NTxOld_meta,
                          "Age",
                          "Gender",
                          "Years_since_transplantation",
                          "BMI",
                          "BSA",
                          "BIAWAISTHIP",
                          "BIAFATPC",
                          "SBP",
                          "Proteinuria",
                          "Retransplantation",
                          "func_constipation",
                          "IBS",
                          "func_diarrhea",
                          "Hypertension",
                          "Alcohol",
                          "Smoking",
                          "Diabetes",
                          "GGT",
                          "ALT",
                          "AST",
                          "ALP",
                          "Tbil",
                          "CRP",
                          "Leukocytes",
                          "eGFR",
                          "CICLOVBLD",
                          "MYCO_PLAS",
                          "TACRO_VBLD",
                          "HDL",
                          "LDL",
                          "Total_cholesterol",
                          "NT_proBNP",
                          "Troponin_T",
                          "PPI",
                          "Statins",
                          "Laxative",
                          "Antibiotics",
                          "Mycophenolic_acid",
                          "Azathioprine",
                          "Ciclosporine",
                          "Tacrolimus",
                          "Prednisolone",
                          "Sirolimus",
                          "Tac_Pred",
                          "Cyclo_Pred",
                          "MMF_Cyclo",
                          "MMF_Pred",
                          "MMF_Tac",
                          "MMF_Cyclo_Pred",
                          "MMF_Tac_Pred",
                          "Aza_Tac_Pred")

#ADONIS
rownames(adonisVarsTouse) <- rownames(ADONIS_Within_NTxOld_meta)
# do univariate adonis
adonis_meta <- matrix(ncol = 7, nrow=ncol(adonisVarsTouse))
ADONIS_Within_NTxOld_meta$ID <- ADONIS_Within_NTxOld_meta$ID.x
ADONIS_Within_NTxOld_meta$ID.x <- NULL
ADONIS_Within_NTxOld_meta$ID.y <- NULL
rownames(ADONIS_Within_NTxOld_meta) <- ADONIS_Within_NTxOld_meta$Row.names
ADONIS_Within_NTxOld_meta$Row.names <- NULL
for (i in 1:ncol(adonisVarsTouse)){
  varToUse <- colnames(adonisVarsTouse)[[i]]
  print (paste(' >> grinding ',colnames(adonisVarsTouse)[[i]]))
  # make new DF with selected phenotype and taxa 
  tmpDF <- merge.data.frame(ADONIS_Within_NTxOld_meta[,colnames(ADONIS_Within_NTxOld_meta) %in% c("ID",varToUse)] ,
                            spDF,by.x="ID",by.y = "row.names")
  tmpDF$ID.y <- NULL
  tmpDF$ID.x <- NULL
  # get rid of ID (used only for merging)
  tmpDF$ID <- NULL
  # get rid of rows with NAs
  print('debug, DF size before removing NAs:')
  print(dim(tmpDF))
  tmpDF <- tmpDF[complete.cases(tmpDF),]
  print('debug, DF size after removing NAs:')
  print(dim(tmpDF))
  if (ncol(tmpDF) > 1 & nrow(tmpDF) > 9) {
    # calculate BC matrix
    tmpDFforBC <- tmpDF
    tmpDFforBC[[varToUse]] <- NULL
    print('calculating BC matrix')
    bcMat <- vegdist(tmpDFforBC,method = "bray")
    print(paste0('doing adonis for ',varToUse))
    ad<-adonis(bcMat ~ adonisVarsTouse[,i],permutations=999)
    aov_table <- ad$aov.tab
    #Df
    adonis_meta[i,1]=aov_table[1,1]
    #SumsOfSqs
    adonis_meta[i,2]=aov_table[1,2]
    #MeanSqs
    adonis_meta[i,3]=aov_table[1,3]
    #FModel
    adonis_meta[i,4]=aov_table[1,4]
    #R2
    adonis_meta[i,5]=aov_table[1,5]
    #Pval
    adonis_meta[i,6]=aov_table[1,6]
  }
}

adonis_meta= as.data.frame(adonis_meta)
adonis_meta$V7=p.adjust(adonis_meta$V6, method = "BH")
adonis_meta$V8="No"
adonis_meta$V8[adonis_meta$V7<0.05]="Yes"
colnames(adonis_meta)=c("DF", "SumsOfSqs", "MeanSqs", "FModel","R2","pval","FDR(BH)","Significant")
rownames(adonis_meta) = colnames(adonisVarsTouse)
```
```{r}
# View(adonis_meta)
g <- ggplot(adonis_meta, aes(reorder(row.names(adonis_meta), R2), R2, fill=Significant)) + 
  geom_bar(stat = "identity") + coord_flip() + theme_bw() + 
  ylab ("Explained variance (R^2) ") + xlab ("Factor")  + theme(text = element_text(size=11))
plot(g)
```